src/bucket.js

```diff
@@ -2238,22 +2253,30 @@ Bucket.prototype.upload = function(localPath, options, callback) {
     });
   } else {
     // Resort to using the name of the incoming file.
-    newFile = this.file(path.basename(localPath), {
```

Comment:
Let's actually stick with `path.basename` for both cases. That way, it will turn "http://www.example.com/my/custom/file.jpg" into just "file.jpg" instead of "my/custom/file.jpg".